actgen/utils.py

```diff
@@ -4,12 +4,105 @@
 import logging
 from pydoc import locate
 
+import gym
+import argparse
 import torch
-from matplotlib import pyplot as plt
+
+from . import wrappers as wrap
 
 Experience = namedtuple('Experience', ['state', 'action', 'reward', 'next_state', 'done'])
 
 
+class Trial:
+    """
+    a base trial class for running experiments
+    """
+    def __init__(self):
+        pass
+
+    def parse_args(self):
+        parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+        # yapf: disable
+        # common args
+        parser.add_argument('--env_name', type=str, default='CartPole-v0',
+                            choices=['CartPole-v0', 'Pendulum-v0', 'LunarLander-v2'],
+                            help='Which gym environment to use')
+        parser.add_argument('--agent', type=str, default='dqn',
+                            choices=['dqn', 'random', 'action_dqn'],
+                            help='Which agent to use')
+        parser.add_argument('--duplicate', '-d', type=int, default=5,
+                            help='Number of times to duplicate actions')
+        parser.add_argument('--random_actions', default=False, action='store_true',
+                            help='Make the duplicate actions all random actions')
+        parser.add_argument('--seed', '-s', type=int, default=0,
+                            help='Random seed')
+        parser.add_argument('--hyperparams', type=str, default='hyperparams/defaults.csv',
+                            help='Path to hyperparameters csv file')
+        parser.add_argument('--test', default=False, action='store_true',
+                            help='Enable test mode for quickly checking configuration works')
+        parser.add_argument('--results_dir', type=str, default='./results/',
+                            help='Path to the result directory to save model files')
+        parser.add_argument('--tag', type=str, default='default_exp',
+                            help='A tag for the current experiment, used as a subdirectory name for saving models')
+        parser.add_argument('--disable_gpu', default=False, action='store_true',
+                            help='enforce training on CPU')
+
+        # train args
+        parser.add_argument('--regularization', type=str, default='None',
+                            choices=['None', 'l1', 'l2', 'dropout'],
+                            help='what regularization method to use during training')
+        parser.add_argument('--gscore', default=False, action='store_true',
+                            help='Calculate the g-score vs time as training proceeds')
+        parser.add_argument('--oracle', default=False, action='store_true',
+                            help='to perform oracle action generalization')
+
+        # manipulation args
```

Comment:
One strategy I've used in the past for handling args that only apply in certain situations is to have a common argument parser (like the one for `Trial`) and then have that be the parent of a specialized parser:

```python
specialized_parser = argparse.ArgumentParser(
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
        parents=[common_parser],
)
```

Reply:
ah ok thanks for the suggestion, this is a neat trick

Reply:
I assume each specialized parser should all have the code to parse unknown args. That seems to results in duplicate code. (Alternatively, we could make the code to parse unknown args into a common function called by each specialized parse, but that might make the code harder to read because it's more spread out). So I was thinking maybe we should keep this code block as it is?

Reply:
I'm totally fine with the other args functionality being abstracted into a `utils` function

Reply:
I'd prefer to have only the args that apply for each script, and ideally we can find a way to avoid having to maintain duplicate arguments in every script